src/types/typechecker.rs

```diff
@@ -1734,7 +1734,7 @@ impl<'a> Inferable<'a> for ast::If<'a> {
             unify(&then.typ, &otherwise.typ, self.location, cache, msg);
             result.with_type(then.typ)
         } else {
-            result.with_type(Type::UNIT)
+            result.with_type(Type::UserDefined(MAYBE_TYPE))
```

Comment:
Looks like you're returning a Maybe type constructor here without applying it to the relevant type arguments. E.g. it is like you're returning `Ptr` instead of `Ptr Int`. You'll likely want to return a TypeApplication of Maybe and then.typ.

We also may need codegen support for this as well....
Perhaps an easier approach for now may be to desugar `if a then b` into `if a then Some b else None` instead.